File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/oauth/OAuthManager.java
package kore.botssdk.audiocodes.oauth;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import org.json.JSONObject;

import java.net.HttpURLConnection;

import kore.botssdk.utils.AsyncTasks;

public class OAuthManager {

    protected static final String TAG = "OAUTH";

    private boolean active;

    private EventsCallback eventsCallback;

    private static SharedPreferences sharedPreferences;
    private static SharedPreferences.Editor editorSP;
    private static final String PREFS_KEY = "oauth_prefs";
    private static final String REFRESH_TOKEN_KEY = "refreshToken";
    private static final String ACCESS_TOKEN_KEY = "accessToken";
    private static final String URL_KEY = "url";
    private static final String REALM_KEY = "realm";
    private static final String CLINET_ID_KEY = "clientId";

    private static OAuthManager instance;

    public static synchronized OAuthManager getInstance() {
        if (instance == null) {
            instance = new OAuthManager();
        }
        return instance;
    }

    public boolean isEnabled() {
        return getValue(ACCESS_TOKEN_KEY) != null;
    }

    public void initialize(Context context, EventsCallback eventsCallback) {
        Log.d(TAG, "initialize");
        this.eventsCallback = eventsCallback;
        sharedPreferences = context.getSharedPreferences(PREFS_KEY, Context.MODE_PRIVATE);
        editorSP = sharedPreferences.edit();
        checkToken(context);
    }

    public void setClientId(String clientId) {
        putValue(CLINET_ID_KEY, clientId);
    }

    public void setURL(String url) {
        putValue(URL_KEY, url);
    }

    public void setRealm(String realm) {
        putValue(REALM_KEY, realm);
    }

    public void authorize(Context context, String userName, String password, LoginCallback loginCallback) {
        if (userName != null && password != null) {
            putValue(REFRESH_TOKEN_KEY, null);
            Log.d(TAG, "authorize");
            new AuthorizeTask(context, userName, password, loginCallback).executeAsync();
        } else {
            Log.d(TAG, "userName || password null");
        }
    }

    private class AuthorizeTask extends AsyncTasks<Void, Void, Void> {
        private final String userName;
        private final String password;
        private final LoginCallback loginCallback;
        private JSONObject jsonObject;
        private final Context context;

        public AuthorizeTask(Context context, String userName, String password, LoginCallback loginCallback) {
            this.context = context;
            this.userName = userName;
            this.password = password;
            this.loginCallback = loginCallback;
        }

        @Override
        protected void onPostExecute() {
            handleResults(context, jsonObject, loginCallback);
        }

        @Override
        protected void doInBackground(Void... arg0) {
            try {
                HttpURLConnection connection = HttpManager.getConnection(makeURL());
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
                String body = "grant_type=" + "password" +
                        "&username=" + userName +
                        "&password=" + password +
                        "&client_id=" + getValue(CLINET_ID_KEY);
                HttpManager.setBody(connection, body);
                jsonObject = HttpManager.make(connection);
            } catch (Throwable e) {
                Log.e(TAG, "AuthorizeTask", e);
            }
        }
    }

    private class RefreshTokenTask extends AsyncTasks<Void, Void, Void> {
        private JSONObject jsonObject;
        private final Context context;

        public RefreshTokenTask(Context context) {
            this.context = context;
        }

        @Override
        protected void onPostExecute() {
            handleResults(context, jsonObject, null);
        }

        @Override
        protected void doInBackground(Void... arg0) {
            try {
                HttpURLConnection connection = HttpManager.getConnection(makeURL());
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
                String body = "grant_type=" + "refresh_token" + "&refresh_token=" + getValue(REFRESH_TOKEN_KEY) +
                        "&client_id=" + getValue(CLINET_ID_KEY);
                HttpManager.setBody(connection, body);
                jsonObject = HttpManager.make(connection);
            } catch (Throwable e) {
                Log.e(TAG, "RefreshTokenTask", e);
            }
        }
    }

    private String makeURL() {
        String server = getValue(URL_KEY);
        String realm = getValue(REALM_KEY);
        return server + "realms/" + realm + "/protocol/openid-connect/token";
    }

    private void handleResults(Context context, JSONObject jsonObject, LoginCallback loginCallback) {
        try {
            if (jsonObject != null && jsonObject.has("access_token")) {
                Log.d(TAG, "success!");
                String access_token = jsonObject.getString("access_token");
                String expires_in = jsonObject.getString("expires_in");
                String refresh_token = jsonObject.getString("refresh_token");
                putValue(ACCESS_TOKEN_KEY, access_token);
                putValue(REFRESH_TOKEN_KEY, refresh_token);
                long tokenExpires = Long.parseLong(expires_in);
                setRefreshTokenAlarm(context, tokenExpires);
                active = true;
                eventsCallback.onUpdateToken(access_token);
                if (loginCallback != null) {
                    loginCallback.onAuthorize(true);
                }
            } else {
                Log.d(TAG, "fail!");
                if (loginCallback != null) {
                    loginCallback.onAuthorize(false);
                } else {
                    putValue(REFRESH_TOKEN_KEY, null);
                    eventsCallback.onRelogin();
                }
            }
        } catch (Throwable e) {
            Log.e(TAG, "Error", e);
            if (loginCallback != null) {
                loginCallback.onAuthorize(false);
            }
        }
    }

    protected void refreshToke(Context context) {
        Log.d(TAG, "refreshToke");
        new RefreshTokenTask(context).executeAsync();
    }

    private void checkToken(Context context) {
        if (isEnabled()) {
            // eventsCallback.onUpdateToken(getValue(ACCESS_TOKEN_KEY));
            Log.d(TAG, "token available");
            if (!active) {
                refreshToke(context);
            }
        } else {
            Log.d(TAG, "no token available");
        }
    }

    protected void setRefreshTokenAlarm(Context context, long refreshTokenExpries) {
        long time = (long) (refreshTokenExpries * 0.9);
        OAuthReceiver.setRefreshTokenAlarm(context, time);
    }

    protected void setAccessTokenAlarm(Context context, long accessTokenExpries) {
        long time = (long) (accessTokenExpries * 0.9);
        OAuthReceiver.setRefreshTokenAlarm(context, time);
    }

    private String getValue(String key) {
        key = key.toLowerCase();
        String value = sharedPreferences.getString(key, null);
        return value;
    }

    private void putValue(String key, String value) {
        key = key.toLowerCase();
        editorSP.putString(key, value);
        editorSP.commit();
    }

    public interface LoginCallback {
        public void onAuthorize(boolean success);
    }

    public interface EventsCallback {
        public void onRelogin();

        public void onUpdateToken(String token);
    }

    public String getURL() {
        return getValue(URL_KEY);
    }

    public String getRealm() {
        return getValue(REALM_KEY);
    }

    public String getClientId() {
        return getValue(CLINET_ID_KEY);
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/oauth/OAuthManager.kt
package com.kore.ui.audiocodes.oauth

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.os.AsyncTask
import com.kore.common.utils.LogUtils
import com.kore.ui.audiocodes.oauth.HttpManager.getConnection
import com.kore.ui.audiocodes.oauth.HttpManager.make
import com.kore.ui.audiocodes.oauth.HttpManager.setBody
import org.json.JSONObject
import java.util.Locale

class OAuthManager private constructor() {
    private var active = false
    private var eventsCallback: EventsCallback? = null
    private lateinit var context: Context
    val isEnabled: Boolean
        get() = getValue(ACCESS_TOKEN_KEY) != null

    fun initialize(context: Context, eventsCallback: EventsCallback?) {
        this.context = context
        LogUtils.d(TAG, "initialize")
        this.eventsCallback = eventsCallback
        sharedPreferences = context.getSharedPreferences(PREFS_KEY, Context.MODE_PRIVATE).apply { editorSP = edit() }
        checkToken()
    }

    fun setUrl(url: String?) {
        putValue(URL_KEY, url)
    }

    fun authorize(context: Context, userName: String?, password: String?, loginCallback: LoginCallback?) {
        this.context = context
        if (userName != null && password != null) {
            putValue(REFRESH_TOKEN_KEY, null)
            LogUtils.d(TAG, "authorize")
            AuthorizeTask(userName, password, loginCallback).execute()
        } else {
            LogUtils.d(TAG, "userName || password null")
        }
    }

    private inner class AuthorizeTask(
        private val userName: String,
        private val password: String,
        private val loginCallback: LoginCallback?
    ) : AsyncTask<Void?, Void?, Void?>() {
        private var jsonObject: JSONObject? = null
        override fun onPostExecute(result: Void?) {
            super.onPostExecute(result)
            handleResults(jsonObject, loginCallback)
        }

        override fun doInBackground(vararg p0: Void?): Void? {
            try {
                val connection = getConnection(makeURL())
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                val body = StringBuilder()
                body.append("grant_type=").append("password")
                    .append("&username=").append(userName)
                    .append("&password=").append(password)
                    .append("&client_id=").append(getValue(CLIENT_ID_KEY))
                setBody(connection, body.toString())
                jsonObject = make(connection)
            } catch (e: Throwable) {
                LogUtils.e(TAG, "AuthorizeTask$e")
            }
            return null
        }
    }

    private inner class RefreshTokenTask : AsyncTask<Void?, Void?, Void?>() {
        private var jsonObject: JSONObject? = null
        override fun onPostExecute(result: Void?) {
            super.onPostExecute(result)
            handleResults(jsonObject, null)
        }

        override fun doInBackground(vararg p0: Void?): Void? {
            try {
                val connection = getConnection(makeURL())
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                val body = StringBuilder()
                body.append("grant_type=").append("refresh_token")
                    .append("&refresh_token=").append(getValue(REFRESH_TOKEN_KEY))
                    .append("&client_id=").append(getValue(CLIENT_ID_KEY))
                setBody(connection, body.toString())
                jsonObject = make(connection)
            } catch (e: Throwable) {
                LogUtils.e(TAG, "RefreshTokenTask $e")
            }
            return null
        }
    }

    private fun makeURL(): String {
        val server = getValue(URL_KEY)
        val realm = getValue(REALM_KEY)
        val url = StringBuilder()
        url.append(server).append("realms/").append(realm).append("/protocol/openid-connect/token")
        return url.toString()
    }

    private fun handleResults(jsonObject: JSONObject?, loginCallback: LoginCallback?) {
        try {
            if (jsonObject != null && jsonObject.has("access_token")) {
                LogUtils.d(TAG, "success!")
                val accessToken = jsonObject.getString("access_token")
                val expiresIn = jsonObject.getString("expires_in")
                val refreshToken = jsonObject.getString("refresh_token")
                putValue(ACCESS_TOKEN_KEY, accessToken)
                putValue(REFRESH_TOKEN_KEY, refreshToken)
                val tokenExpires = expiresIn.toLong()
                setRefreshTokenAlarm(context, tokenExpires)
                active = true
                eventsCallback?.onUpdateToken(accessToken)
                loginCallback?.onAuthorize(true)
            } else {
                LogUtils.d(TAG, "fail!")
                if (loginCallback != null) {
                    loginCallback.onAuthorize(false)
                } else {
                    putValue(REFRESH_TOKEN_KEY, null)
                    eventsCallback?.onReLogin()
                }
            }
        } catch (e: Throwable) {
            LogUtils.e(TAG, "Error $e")
            loginCallback?.onAuthorize(false)
        }
    }

    fun refreshToken() {
        LogUtils.d(TAG, "refreshToke")
        RefreshTokenTask().execute()
    }

    private fun checkToken() {
        if (isEnabled) {
            LogUtils.d(TAG, "token available")
            if (!active) refreshToken()
        } else {
            LogUtils.d(TAG, "no token available")
        }
    }

    private fun setRefreshTokenAlarm(context: Context, refreshTokenExpries: Long) {
        val time = (refreshTokenExpries * 0.9).toLong()
        OAuthReceiver.setRefreshTokenAlarm(context, time)
    }

    private fun setAccessTokenAlarm(context: Context, accessTokenExpries: Long) {
        val time = (accessTokenExpries * 0.9).toLong()
        OAuthReceiver.setRefreshTokenAlarm(context, time)
    }

    private fun getValue(key: String): String? {
        var newKey = key
        newKey = newKey.lowercase(Locale.getDefault())
        return sharedPreferences?.getString(newKey, null)
    }

    private fun putValue(key: String, value: String?) {
        var newKey = key
        newKey = newKey.lowercase(Locale.getDefault())
        editorSP?.putString(newKey, value)
        editorSP?.commit()
    }

    interface LoginCallback {
        fun onAuthorize(success: Boolean)
    }

    interface EventsCallback {
        fun onReLogin()
        fun onUpdateToken(token: String?)
    }

    fun setURL(url: String?) {
        putValue(URL_KEY, url)
    }

    fun getURL(): String? {
        return getValue(URL_KEY)
    }

    fun getRealm(): String? {
        return getValue(REALM_KEY)
    }

    fun setRealm(realm: String?) {
        putValue(REALM_KEY, realm)
    }

    fun setClientId(clientId: String?) {
        putValue(CLIENT_ID_KEY, clientId)
    }

    fun getClientId(): String? {
        return getValue(CLIENT_ID_KEY)
    }

    companion object {
        const val TAG = "OAUTH"
        private var sharedPreferences: SharedPreferences? = null
        private var editorSP: SharedPreferences.Editor? = null
        private const val PREFS_KEY = "oauth_prefs"
        private const val REFRESH_TOKEN_KEY = "refreshToken"
        private const val ACCESS_TOKEN_KEY = "accessToken"
        private const val URL_KEY = "url"
        private const val REALM_KEY = "realm"
        private const val CLIENT_ID_KEY = "clientId"
        @SuppressLint("StaticFieldLeak")
        private var instance: OAuthManager? = null

        @Synchronized
        fun getInstance(): OAuthManager {
            if (instance == null) instance = OAuthManager()
            return instance!!
        }

    }
}